q2_sample_classifier/_format.py

```diff
@@ -6,59 +6,86 @@
 # The full license is in the file LICENSE, distributed with this software.
 # ----------------------------------------------------------------------------
 
-import importlib
 import tarfile
 import json
 
 import qiime2.plugin.model as model
 from qiime2.plugin import ValidationError
 
 
+def _validate_record_len(cells, current_line_number, exp_len):
+    if len(cells) != exp_len:
+        raise ValidationError(
+            "Expected data record to be TSV with {0} "
+            "fields. Detected {1} fields at line {2}:\n\n{2!r}"
+            .format(exp_len, len(cells), current_line_number, cells))
+
+
+def _validate_file_not_empty(has_data):
+    if not has_data:
+        raise ValidationError(
+            "There must be at least one data record present in the "
+            "file in addition to the header line.")
+
+
 class BooleanSeriesFormat(model.TextFileFormat):
-    def sniff(self):
+    def _validate_(self, level):
+        n_records = {'min': 5, 'max': None}[level]
         with self.open() as fh:
+            # validate header
+            # for now we will not validate any information in the header.
             line = fh.readline()
-            for line, _ in zip(fh, range(5)):
+
+            # validate body
+            has_data = False
+            for line_number, line in enumerate(fh, start=2):
                 cells = line.strip().split('\t')
-                if len(cells) != 2 or str(cells[1]) not in ('True', 'False'):
-                    return False
-            return True
+                _validate_record_len(cells, line_number, 2)
+                if str(cells[1]) not in ('True', 'False'):
+                    raise ValidationError(
+                        "Expected data to be comprised of values `True` and "
+                        "`False`, found {0} at line {1}."
+                        .format(str(cells[1]), line_number))
+                has_data = True
+                if n_records is not None and (line_number - 1) >= n_records:
+                    break
+
+            _validate_file_not_empty(has_data)
 
 
 BooleanSeriesDirectoryFormat = model.SingleFileDirectoryFormat(
     'BooleanSeriesDirectoryFormat', 'outliers.tsv',
     BooleanSeriesFormat)
 
 
+# This is effectively an internal format - it isn't registered with the
```

Comment:
@ebolyen - this shouldn't be a problem, right? At the very least, it suggests these formats might benefit from being centralized into q2-types, although that probably isn't worth it until we address the "format schema" problem (or at least come up with a plan there).

Reply:
that's definitely weird, but since its working, lets not think too hard about it yet (pending better stuff with schemas)